ft_virtualchannel.m

```diff
@@ -0,0 +1,382 @@
+  
+  nvc = numel(cfg.parcel);
+end
+
+% create a montage for each of the dipoles/parcels, and use ft_apply_montage,
+% followed by a dimensionality reduction step, using ft_componentanalysis
+unmixing  = cell(1, nvc);
+topolabel = cell(1, nvc);
+tmpdata   = cell(1, nvc);
+for i = 1:nvc
+  
+  montage = [];
+  if usepos
+    
+    montage.tra      = source.filter{indx(i)};
+    montage.labelold = source.label;
+    montage.labelnew = cell(size(montage.tra,1), 1);
+    for k = 1:size(montage.tra,1)
+      montage.labelnew{k} = sprintf('virtualchannel%03d_orientation%03d', i, k);
+    end
+    bname = sprintf('virtualchannel%03d', i);
+    
+  elseif useparcellation
+  
+    sel = find(parcellation.(cfg.parcellation)==indx(i));
+    montage.tra      = cat(1, source.filter{sel});
+    montage.labelold = source.label;
+    montage.labelnew = cell(size(montage.tra,1), 1);
+    cnt = 0;
+    for k = 1:numel(sel)
+      for kk = 1:size(source.filter{sel(k)},1)
+        cnt = cnt+1;
+        montage.labelnew{cnt} = sprintf('%s_dipole%03d_orientation%03d', cfg.parcel{i}, k, kk);
+      end
+    end
+    bname = matlab.lang.makeValidName(cfg.parcel{i}); % FIXME not sure how universal this is.
+    
+  end
+  
+  % apply the montage to the numeric data, but make an exception for the
+  % svd method: this one is much more efficient, in that its implementattion
+  % operates on the covariance matrix directly, bypasses ft_componentanalysis, 
+  % and does not require the intermediate (possibly memory greedy, and slow)
+  % data representation.
+  if ~isequal(cfg.method, 'svd')
+    tmpdata{1,i}  = ft_apply_montage(data, montage, 'feedback', 'none');
+    
+    [i1, i2] = match_str(data.label, montage.labelold);
+    unmixing{1,i}(:,i1) = montage.tra(:,i2);
+    topolabel{1,i}      = data.label(i1);
+    
+    % apply the montage to the sensor description
+    sensfields = {'grad' 'elec' 'opto'};
+    for k = 1:numel(sensfields)
+      if isfield(tmpdata{i}, sensfields{k})
+        ft_info(sprintf('applying the montage to the %s structure\n', sensfields{k}));
+        tmpdata{i}.(sensfields{k}) = ft_apply_montage(tmpdata{i}.grad, montage, 'feedback', 'none', 'keepunused', 'no', 'balancename', bname);
+      end
+    end
+    
+    compmethods = {'pca' 'runica' 'fastica' 'dss'};
+    switch cfg.method
+      case compmethods
+        
+        tmpcfg     = keepfields(cfg, {'method' cfg.method 'numcomponent'});
+        if isequal(tmpcfg.numcomponent, 'all')
+          tmpcfg = rmfield(tmpcfg, 'numcomponent');
+        end
+        
+        tmpdata{i}  = ft_componentanalysis(tmpcfg, tmpdata{i});
+        unmixing{i} = tmpdata{i}.unmixing * unmixing{i};
+        
+        for k = 1:numel(tmpdata{i}.label)
+          if usepos
+            str = sprintf('virtualchannel%03d', i);
+          else
+            str = cfg.channel{i};
+          end
+        end
+        tmpdata{i}.label{k} = sprintf('%s_%s%03d', str, cfg.method, k);
```

Comment:
I suspect the most common use case here will be `ncomp == 1`, right? In that case, people might not care for the `runicaXXX` suffix, and just want the parcel (/pos) label. So I propose:

```suggestion
        if cfg.numcomponent == 1
          tmpdata{i}.label{k} = str;
        else
          tmpdata{i}.label{k} = sprintf('%s_%s%03d', str, cfg.method, k);
        end
```

Reply:
well, I guess that my intention would be that it might still be relevant to reflect in the label the heuristic used for the orientation estimation, in which case 'pca'/'dss' etc still are relevant. I agree that the k can be dropped if ncomp=1

Reply:
I'm inclined to say the orientation estimation method does not need to be explicitly reflected in the virtual channel label. (We also don't have a `freq.powspctrm_hanning` or `freq.powspctrm_wavelet`.) In most cases I expect people to be happy/happier with `data.label = {'Occipital_R_4', ...}` than `data.label = {'Occipital_R_4_runica', ...}`. But, having said that, I don't have any strong objections either ;)

Note, by the way, I think you can accept my suggestions directly in GitHub (and then pull them locally), so you don't have to write them again and push (like you did with the typo).

Reply:
That's a fair argument. And I can imagine that a person who cares about this, would probably know how to extract the information (assuming that all goes well with the cfg tracking...)

Reply:
and yes, I of course could just have accepted the typo fixes, didn't think of that. Thanks.